File: 0505_the_maze_II/shortest_distance.py
class Solution:
    def shortestDistance(self, maze: List[List[int]], start: List[int], destination: List[int]) -> int:
        heap = [[0] + start]
        dists = defaultdict(lambda: float('inf'))
        dists[tuple(start)] = 0
        while heap:
            distance, i, j = heapq.heappop(heap)
            if [i, j] == destination:
                return distance
            for x, y in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                r, c, count = i, j, distance
                while len(maze) > r + x >= 0 and len(maze[0]) > c + y >= 0 and maze[r + x][c + y] == 0:
                    r, c, count = r + x, c + y, count + 1
                if count < dists[(r, c)]:
                    heapq.heappush(heap, (count, r, c))
                    dists[(r, c)] = count
        return -1

File: 0505_the_maze_II/shortest_distance.java
public class Solution {
    // Time: O(n * m * log(n * m))
    // Space: O(n * m)
    public int shortestDistance(int[][] maze, int[] start, int[] dest) {
        int[][] distance = new int[maze.length][maze[0].length];
        for (int[] row : distance)
            Arrays.fill(row, Integer.MAX_VALUE);
        distance[start[0]][start[1]] = 0;
        dijkstra(maze, start, distance);
        int finalDest = distance[dest[0]][dest[1]];
        return finalDest == Integer.MAX_VALUE ? -1 : finalDest;
    }

    public void dijkstra(int[][] maze, int[] start, int[][] distance) {
        int[][] dirs = { { 0, 1 }, { 0, -1 }, { -1, 0 }, { 1, 0 } };
        PriorityQueue<int[]> queue = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        queue.offer(new int[] { start[0], start[1], 0 });
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            if (distance[curr[0]][curr[1]] < curr[2])
                continue;
            for (int[] dir : dirs) {
                int x = dir[0], y = dir[1], count = 0;
                int r = curr[0], c = curr[1];
                while (r + x >= 0 && c + y >= 0 && r + x < maze.length && c + y < maze[0].length
                        && maze[r + x][c + y] == 0) {
                    r += x;
                    c += y;
                    count++;
                }
                int curr_dist = distance[curr[0]][curr[1]] + count;
                if (curr_dist < distance[r][c]) {
                    distance[r][c] = curr_dist;
                    queue.offer(new int[] { r, c, curr_dist });
                }
            }
        }
    }
}